### DISCUSSION TITLE
Function executed by job_queue doesn't send Inline Keyboard to a user.

### DISCUSSION THREAD

**Discussion by cerebrumdk** [python-telegram-bot/python-telegram-bot#2616] [Q&A] [answered]
Hello, I didn't find answer to this question (and the question itself) on the web, so decided to ask you. I hope someone will help me, it would be greatly appreciated. 
I'm writing schedule bot, kind of. Bot accepts user's schedule, transforms it into nested dictionary and saves to user_data 
Example of dictionary:
{'Schedule': 
    {'1)Monday': 
        {'7:00': 'Wake up',
         '8:00': 'Work',
         '22:00': 'Trying not to die'}
    }
}
and so on. 
User is able to set reminders (or rather "checkers") for some day or the whole week:
/set_reminders 1)Monday
Each key in context.user_data['Schedule']['1)Monday'] will become a timestamp for job_queue.
The idea is that user will got a message at, for example 7:00 saying:
'Are you prepared/already doing this task: "Wake up"?'
/Three buttons/
(Yes) (No)
(No(Circumstances)) 
**My problem is that keyboard doesn't get sent.**

### I've created two minimal working examples:

**This one is way shorter**, I think it's better to check that one, since it doesn't have set_schedule function, we simply change time and day in the dictionary that I've created manually. Works just the same, error is just the same, but it's half the size of the second one.
https://hastebin.com/wulalakica.py
![image](https://user-images.githubusercontent.com/83335108/128424221-125f7710-1af5-448a-a080-87c89acfa5e1.png)

Second one is a full implementation of set_reminders where we set not only reminders but also schedule dictionary through telegram and it gets stored. I've added screenshot of "user steps" 
https://hastebin.com/etevemokex.py
![image](https://user-images.githubusercontent.com/83335108/128422229-6e8f5d0e-0757-402a-9f8f-e81185514779.png)

Here's a snippet of my code and an error that I get with short explanations:
1) **Function with job_queue** (I'm not sure either you need the whole function, so I'm adding additional print screen with just run_daily function:

```
 def set_reminders(update, context):
    reminder_list = context.args
    reminder = ''.join(reminder_list) # It always will be one word: either 'all' or f'day'
    schedule = context.user_data['Schedule']
    if reminder == 'all':
        for day in schedule.keys():
            j_day_form = int(day[0]) - 1 # changing format of a day, because 
                                         # days in run_daily starts from 0 and includes only numbers.  
            for timestamp in schedule[day].keys():
                r_name = str(j_day_form) + ')' + timestamp
                r_hour, r_minute =time_formatting(timestamp)
                print(r_minute, r_hour)
                context.job_queue.run_daily(in_time, 
                                            time(hour=r_hour, minute=r_minute, tzinfo=pytz.timezone('Europe/Kiev')),
                                            days=([j_day_form]), 
                                            context=[update.message.chat_id, schedule[day][timestamp]], 
                                            name=r_name)
    else:
        j_day_form = int(reminder[0]) - 1
        for timestamp in schedule[reminder].keys():
            r_name = str(j_day_form) + ')' + timestamp
            r_hour, r_minute = time_formatting(timestamp)
            context.job_queue.run_daily(in_time, 
                                        time(hour=r_hour, minute=r_minute, tzinfo=pytz.timezone('Europe/Kiev')),
                                        days=([j_day_form]), 
                                        context=[update.message.chat_id, schedule[reminder][timestamp]], 
                                        name=r_name)
```

(screenshot1)

run_daily only:

```
 context.job_queue.run_daily(in_time, 
                                        time(hour=r_hour, minute=r_minute, tzinfo=pytz.timezone('Europe/Kiev')),
                                        days=([j_day_form]), 
                                        context=[update.message.chat_id, schedule[reminder][timestamp]], 
                                        name=r_name)
```

(screenshot2)

2) **Function that job_queue executes**: 

```
 def in_time(context):
    in_time_keyboard = [
        [InlineKeyboardButton('Yes', callback_data='Yes'),
        InlineKeyboardButton('No', callcack_data='No'),
        InlineKeyboardButton('No (Circumstances)', callback_data='C')]
    ]

    reminder_reply_markup = InlineKeyboardMarkup(in_time_keyboard)
    context.bot.send_message(chat_id=context.job.context[0], 
                            text=f'Are you doing/prepared to do this task?\n{context.job.context[1]}')
                            
    context.bot.send_message(chat_id=context.job.context[0], reply_markup=reminder_reply_markup)
```

(screenshot3)

 3) **Error**. It's so long that I'm not sure either it has any valuable info:D
It executes everything just fine, it adds job and runs it as well, as you can see at the very top. If I delete line context.bot.send_message(chat_id=context.job.context[0], reply_markup=reminder_reply_markup)  in screenshot 3, then no error is being raised (It executes everything just the same  **with** that line as well, but raises error and doesn't show buttons). So, in both variants user has a message but no buttons to click

`Message: 'Added job "%s" to job store "%s"'
Arguments: ('2)23:24', 'default')     =====> **As u can see, the job is added, but when time comes, keyboard doesn't show itself.**
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/__init__.py", line 440, in format
    return self._format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 436, in _format
    return self._fmt % record.__dict__
KeyError: 'asci'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/__init__.py", line 1085, in emit
    msg = self.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 929, in format
    return fmt.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 671, in format
    s = self.formatMessage(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 640, in formatMessage
    return self._style.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 442, in format
    raise ValueError('Formatting field not found in record: %s' % e)
ValueError: Formatting field not found in record: 'asci'
Call stack:
  File "/usr/lib/python3.8/threading.py", line 890, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 80, in _worker
    work_item.run()
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.8/dist-packages/apscheduler/executors/base.py", line 123, in run_job
    logger.info('Running job "%s" (scheduled at %s)', job, run_time)
Message: 'Running job "%s" (scheduled at %s)'
Arguments: (<Job (id=776bde33ec1c4e69b74a788986d17e0c name=2)23:24)>, datetime.datetime(2021, 8, 4, 23, 24, tzinfo=<DstTzInfo 'Europe/Kiev' EEST+3:00:00 DST>))
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/__init__.py", line 440, in format
    return self._format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 436, in _format
    return self._fmt % record.__dict__
KeyError: 'asci'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/__init__.py", line 1085, in emit
    msg = self.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 929, in format
    return fmt.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 671, in format
    s = self.formatMessage(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 640, in formatMessage
    return self._style.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 442, in format
    raise ValueError('Formatting field not found in record: %s' % e)
ValueError: Formatting field not found in record: 'asci'
Call stack:
  File "/usr/lib/python3.8/threading.py", line 890, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 80, in _worker
    work_item.run()
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 63, in run
    self.future.set_result(result)
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 538, in set_result
    self._invoke_callbacks()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 328, in _invoke_callbacks
    callback(self)
  File "/usr/local/lib/python3.8/dist-packages/apscheduler/executors/pool.py", line 20, in callback
    self._run_job_success(job.id, f.result())
  File "/usr/local/lib/python3.8/dist-packages/apscheduler/executors/base.py", line 90, in _run_job_success
    self._scheduler._dispatch_event(event)
  File "/usr/local/lib/python3.8/dist-packages/apscheduler/schedulers/base.py", line 831, in _dispatch_event
    cb(event)
  File "/usr/local/lib/python3.8/dist-packages/telegram/ext/jobqueue.py", line 86, in _dispatch_error
    self._dispatcher.dispatch_error(None, event.exception)
  File "/usr/local/lib/python3.8/dist-packages/telegram/ext/dispatcher.py", line 818, in dispatch_error
    self.logger.exception(
Message: 'No error handlers are registered, logging exception.'
Arguments: ()`

(screenshot4-6)

Screenshots:
screenshot1
![image](https://user-images.githubusercontent.com/83335108/128002163-23969a85-993c-4a0e-b6aa-289ca22b75ce.png)
screenshot2
![image](https://user-images.githubusercontent.com/83335108/128002336-c02b12f2-28d2-488d-8a66-1546ffb28167.png)
screenshot3
![image](https://user-images.githubusercontent.com/83335108/128003237-911a6439-a9ac-48df-9bfc-0d6cfdce6d5b.png)
screenshot4
![image](https://user-images.githubusercontent.com/83335108/128006589-579c1c88-f675-4b1c-a4ad-6bfd3c8a3ca9.png)
screenshot5
![image](https://user-images.githubusercontent.com/83335108/128006644-2e0bafa0-dcb8-43f2-9e74-6b3291acd927.png)
screenshot6
![image](https://user-images.githubusercontent.com/83335108/128006690-fd1278c6-6008-4de1-9da8-6c22beb9b7e7.png)

**Comment by Poolitzer** [comment]
Could you include the code blocks and the traceback as copy pasted code blocks? That would make it so much easier to read.

**Reply by cerebrumdk** [reply]
Thanks for suggestion, I've just edited my question with code blocks:)
Sorry for the delay though, couldn't do it earlier.

**Comment by Poolitzer** [comment]
Alright, thanks for the update. Looking at that code, I can't see the error. It would be great if you can make a working snippet out of that so I can run it on my end, it will be way easier for me to debug. https://telegra.ph/Minimal-Working-Example-for-PTB-07-18 is a good overview for that.

Also, your example dict is wrong, it should be like: `{'Schedule': {'1)Monday': {'7:00': 'Wake up', '8:00': 'Work', '22:00': 'Trying not to die'}}}`. At least I assume. But that would give you a very different error I am sure.

**Reply by cerebrumdk** [reply]
Updated with a minimal working example, hope I did it correctly. 
About dictionary example: yes, my mistake, bot stores schedule as you wrote it. I've updated dictionary example as well.

**Comment by Poolitzer** [comment]
Aha. I found the issue, it in def in_time(context)


```
context.bot.send_message(chat_id=context.job.context[0], 
                            text=f'Are you doing/prepared to do this task?\n{context.job.context[1]}')
                            
context.bot.send_message(chat_id=context.job.context[0], reply_markup=reminder_reply_markup)

```

A new message needs to have a text, or you need to attach the reply_markup to the first message. So either

 ```
context.bot.send_message(chat_id=context.job.context[0], 
                            text=f'Are you doing/prepared to do this task?\n{context.job.context[1]}')
                            
context.bot.send_message(chat_id=context.job.context[0], text="Look at me I am a text", reply_markup=reminder_reply_markup)

```

or

```
context.bot.send_message(chat_id=context.job.context[0], 
                            text=f'Are you doing/prepared to do this task?\n{context.job.context[1]}', reply_markup=reminder_reply_markup)

```

**Reply by cerebrumdk** [reply]
I've tried both methods, they don't work. If I use second one, then message doesn't get sent at all

**Reply by Poolitzer** [reply]
damn it

**Comment by Poolitzer** [comment]
Since I apparently didnt find the issues, I wanted to try your working examples, but both links just give me a blank webpage.

Could you upload the source code somewhere else? What always works for me is https://gist.github.com or https://pastebin.com

**Reply by cerebrumdk** [reply]
https://pastebin.com/2Mux40T0
Timezones: 'Etc/GMT', 'Etc/GMT+0', 'Etc/GMT+1', 'Etc/GMT+2', 'Etc/GMT+3', 'Etc/GMT+4',
                 'Etc/GMT+5', 'Etc/GMT+6', 'Etc/GMT+7', 'Etc/GMT+8', 'Etc/GMT+9',
                 'Etc/GMT+10', 'Etc/GMT+11', 'Etc/GMT+12', 'Etc/GMT-0',
                 'Etc/GMT-1', 'Etc/GMT-2', 'Etc/GMT-3', 'Etc/GMT-4', 'Etc/GMT-5', 'Etc/GMT-6', 'Etc/GMT-7',
                 'Etc/GMT-8', 'Etc/GMT-9', 'Etc/GMT-10', 'Etc/GMT-11', 'Etc/GMT-12', 'Etc/GMT-13', 'Etc/GMT-14',
                 'Etc/GMT0', 'Etc/Greenwich', 'Etc/UCT', 'Etc/UTC', 'Etc/Universal', 'EET', 'EST', 'EST5EDT'.

**Comment by Poolitzer** [comment, marked as answer]
JESUS CHRIST I FOUND IT.

Two issues which made this so hard for both of us to hunt down:

1. Your logger was broken. That way, you weren't able to see the actually error which triggered the long traceback you encountered. Once I was faced with it as well, I found the issue:

`logging.basicConfig(format='%(asci)s - %(name)s - %(levelname)s - %(message)s',
                     level=logging.INFO)`
is wrong. One correct way would be
 `logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                    level=logging.INFO)`.

If you set `asci` there intentionally, you have to rethink why, because its not correct :D

Once that was fixed and I changed the code a bit so in_time would be called, the error I faced was:

```
2021-08-15 20:46:00,129 - telegram.ext.dispatcher - ERROR - No error handlers are registered, logging exception.
Traceback (most recent call last):
  File "C:\Users\Poolitzer\PycharmProjects\ReportBot\venv\lib\site-packages\apscheduler\executors\base.py", line 125, in run_job
    retval = job.func(*job.args, **job.kwargs)
  File "C:/Users/Poolitzer/AppData/Roaming/JetBrains/PyCharm2021.2/scratches/scratch_1.py", line 32, in in_time
    context.bot.send_message(chat_id=context.job.context[0],
  File "C:\Users\Poolitzer\PycharmProjects\ReportBot\venv\lib\site-packages\telegram\bot.py", line 130, in decorator
    result = func(*args, **kwargs)
  File "C:\Users\Poolitzer\PycharmProjects\ReportBot\venv\lib\site-packages\telegram\bot.py", line 513, in send_message
    return self._message(  # type: ignore[return-value]
  File "C:\Users\Poolitzer\PycharmProjects\ReportBot\venv\lib\site-packages\telegram\ext\extbot.py", line 199, in _message
    result = super()._message(
  File "C:\Users\Poolitzer\PycharmProjects\ReportBot\venv\lib\site-packages\telegram\bot.py", line 332, in _message
    result = self._post(endpoint, data, timeout=timeout, api_kwargs=api_kwargs)
  File "C:\Users\Poolitzer\PycharmProjects\ReportBot\venv\lib\site-packages\telegram\bot.py", line 295, in _post
    return self.request.post(
  File "C:\Users\Poolitzer\PycharmProjects\ReportBot\venv\lib\site-packages\telegram\utils\request.py", line 356, in post
    result = self._request_wrapper(
  File "C:\Users\Poolitzer\PycharmProjects\ReportBot\venv\lib\site-packages\telegram\utils\request.py", line 279, in _request_wrapper
    raise BadRequest(message)
telegram.error.BadRequest: Can't parse inline keyboard button: text buttons are unallowed in the inline keyboard
```
This is still a long traceback, but one I am familiar with. Somewhere in the keyboard definition must have been an issue which only lead to the text argument being recognized, such generating a "text button". After really pinching my eyes together, I found it:
`InlineKeyboardButton('No', callcack_data='No'),`
which has a small typo in it. Which leads to issue number two:

2. The error generation for non expected kwargs in user facing API classes sucks, because there is none. So even if it would have yielded that error, it would have taken me a hot minute to find the mistake in your code, which otherwise is fine. This is what #2607 is for, and this conversation made me firmer in my stance towards it. So thanks for it.


Alright, that's it. It should solve your issues, and I hope you don' have many furthers. I am sorry for taking so long to reply, I have busy weeks behind me.

**Reply by cerebrumdk** [reply]
Sorry for such a late reply. Thank you very much, everything is working now! It's quite symbolic that you answered me on my birthday:D 
Once again, thank you for your time and efforts:)

**Reply by Poolitzer** [reply]
eyy, a birthday present it was :D
